### ISSUE TITLE
Gating function averaging

### ISSUE THREAD

**Issue by justheuristic** [learning-at-home/hivemind#95]
In our preliminary experiments, all peers have independent gating functions and we can only synchronize them manually. It would be great to implement some sort of builtin averaging mechanism.

For instance, every T seconds, assemble peers into groups at random, then perform all-reduce within each group. In case of failure, rollback and repeat T seconds later.

**Comment by justheuristic** [comment]
Based on discussion with @mryab 

Formally speaking, you have a graph ![formula](https://render.githubusercontent.com/render/math?math=G(V,E)) where vertices are participants and edges represent peer relations. Each peer has a vector of parameters ![formula](https://render.githubusercontent.com/render/math?math=v_i) .

We want __all__ peers to converge to the global average 

Using the notation above, consider an (unknown) global average
![formula](https://render.githubusercontent.com/render/math?math=v^*%20%3D%20\sum_{v_i\in%20V}{v_i})

We consider two solutions:
1. each participant periodically sends his gating function to M random peers. When ![formula](https://render.githubusercontent.com/render/math?math=v_j), receives gating function from ![formula](https://render.githubusercontent.com/render/math?math=v_i), he updates his parameters with a moving average ![formula](https://render.githubusercontent.com/render/math?math=v_j%3A%3D{\alpha\cdot%20v_i%2B(1%2D\alpha)\cdot{v_j}})
2. participants ~somehow form groups of up to M members and run butterfly all-reduce between each other

The first approach is simpler to implement but requires O(NM) total communication for N peers, the second approach requires only O(N) communication but it can be tricky to implement with fault tolerance.


Notes:
* Should we send parameters with optimizer stats or just the parameters?

**Comment by justheuristic** [comment]
Based on discussion with @shekhale

Using the notation above, consider an (unknown) global average ![formula](https://render.githubusercontent.com/render/math?math=v^*%20%3D%20\underset{v_i\in%20V}{E}{v_i})

We can postulate the problem of global averaging as an optimization problem with mean squared error objective
![formula](https://render.githubusercontent.com/render/math?math=L%3D\underset{v_i\in%20V}{E}{(v_i-v^*)^2})

The problem with approach 1 (prev post) is that each time one of the peers updates his parameters, the global average changes. As a result, the network will converge to a different global average depending on network structure and the order in which nodes perform averaging.

For approach 2, the AllReduce op for any subset of nodes will preserve the global average

Proof, sketch:
Imagine that two nodes ![formula](https://render.githubusercontent.com/render/math?math={v_i,v_j}) perform allreduce. In this trivial case, ![formula](https://render.githubusercontent.com/render/math?math={v^{new}_i,v^{new}_j%3A%3D\frac{(v_i%2Bv_j)}{2}})

After this transaction, the new global average doesn't change

![formula](https://render.githubusercontent.com/render/math?math=v^*_{new}%20%3D%20{v^*_{old}-v_i-v_j%2Bv^{new}_i%2Bv^{new}_j%3Dv^*_{old}-v_i-v_j%2B2\cdot\frac{(v_i%2Bv_j)}{2}}%3Dv^*_{old})

and the loss function decreases (or stays the same iff ![formula](https://render.githubusercontent.com/render/math?math={v_i%3Dv_j}))

As a result, as long as the graph ![formula](https://render.githubusercontent.com/render/math?math=v^*%20%3D%20\underset{v_i\in%20V}{E}{v_i}) is connected, the procedure will converge to ![formula](https://render.githubusercontent.com/render/math?math={L\rightarrow%200})

The same is true (but harder to tex) for all-reducing in groups of any size.

Notes:
* so far we never relied on the fact that the loss function is a mean __squared__ error, it could just as well have been mean absolute error
* while the procedure converges to the global average, the convergence speed may vary depending on how we select the groups. Heuristics: bigger = better? (because of butterfly allreduce)
* the proof should hold as long as all-reduce is transactional. In practice, we can ensure that by checking that all peers ACK-ed the allreduce before applying changes. Or just ignore failures.
* we can perform a fault-tolerant butterfly AllReduce by ensuring that that each partition is aggregated on not 1 but k peers (k<M)

**Comment by Shekhale** [comment]
How about a hierarchical structure for getting (truly) v*?
Assume we have N = 10**4 nodes, they are all ready to communicate and we can perform (mb fault-tolerant or not) butterfly AllReduce for M = 10 with &delta; ms time cost, this means that we can calculate v*  with  time cost of  &delta; * log<sub>M</sub> N = 4 &delta; after that we can "backward" it for all nodes this the same (logarithmic w.r.t. N) complexity. In other words, instead of investigate algorithm of selecting the "appropriate" nodes and "hoping" or proving that v<sub>i</sub> will converge to v*, we can guarantee this without significantly increasing the time costs.

I believe that "they are all ready to communicate" is not appropriate assumption, but this idea can be modified for the case when nodes become ready to communicate at different time.

**Comment by Maria1007** [comment]
I can not understand what you mean，and I need to learn how create my own Education system online.

**Comment by justheuristic** [comment]
@Maria1007 sanity check: we're talking about training artificial neural networks. If you're after online education, you may have found us by mistake.

**Comment by justheuristic** [comment]
(based on conversations with @mryab @Shekhale )
For the record: here's how we determine the optimal group size and compution for allreduce.

__Intuition:__

Imagine we have N=100 trainers with separate gating functions (parameter vectors). We could average them in one allreduce call, but such allreduce may have too high failure rate (increasing with the number of nodes).

Instead, we could try to split them into 10 groups by 10 nodes. Let's arbitrarily order them in a 10x10 grid.

![image](https://user-images.githubusercontent.com/3491902/97396746-bc988400-18f8-11eb-8dc5-40a420d2e7ca.png)

First, we run allreduce in each (non-overlapping) group of 10 nodes. This results in each of 10 nodes receiving the local average vector over its subgroup (see above).


Then, if we can group nodes "vertically" and perform averaging between these local averages. After that, each node will have __the global average vector__ (see below).

![image](https://user-images.githubusercontent.com/3491902/97397233-d7b7c380-18f9-11eb-844f-8d9038eedf16.png)

We do not have to physically group nodes into horizontal and vertical lines. This technique will work with any distribution as long as __the nodes grouped together in the second round have NOT been grouped together in the first round.


![image](https://user-images.githubusercontent.com/3491902/97397209-cc649800-18f9-11eb-9f15-762d689574a4.png)


This trick can be generalized to an higher-dimensional case: 
![image](https://user-images.githubusercontent.com/3491902/97397598-af7c9480-18fa-11eb-91b9-97cf67d775ce.png)

In the general case, one has to find groups of M nodes that have NOT been used for T previous rounds of allreduce, where T = ceil(log_M {N}) and N is the total number of participants. In practice, since the number of nodes is unlikely to be a power of M, some allreduce groups will end up incomplete.

Since we're only averaging groups of M participants, the probability of success in each group is higher than in case of global allreduce. If some of the groups fail, we only need to restart those specific groups, resulting in a smaller average number of retries.

In other words, large M require less steps but are more likely to fail while small M need more steps but fewer retries. One way to determine the optimal group size M is to maximize the expected time to global average under perfect conditions.

We implemented a simple script that estimates the total time given the total number of participants, chosen group size and failure rate.
[optimal_group_size.zip](https://github.com/learning-at-home/hivemind/files/5449928/optimal_group_size.zip)

Below are 2D plots where
* X axis is the probability of each node failing during one round of allreduce
* Y axis is the total number of nodes
* the value in each cell is the **optimal** group size M chosen by brute force

(main plot)
![image](https://user-images.githubusercontent.com/3491902/97398424-472eb280-18fc-11eb-8271-2f66c6465f5e.png)

(larger total participant counts)
![image](https://user-images.githubusercontent.com/3491902/97398443-51e94780-18fc-11eb-88b6-16ec8eaadb77.png)


We're currently working on a practical decentralized implementation of this averaging algorithm (see PR above).

**Comment by justheuristic** [comment]
Merged in #123 #134  #140
